### PR TITLE
Document serverless limitations

### PR DESCRIPTION
**Pull request by kwiatekus**
**Description**

Changes proposed in this pull request:

- Document serverless limitations

**Related issues**
https://github.com/kyma-project/kyma/issues/12849

### CODE REVIEW

**Review by pPrecel — APPROVED** [review]

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+
+#### Python functions
+
+|                 | local-dev | no profile (no limits for resource) |
+|-----------------|-----------|-------------------------------------|
+| no dependencies | 30 sec    | 16 sec                              |
+| 2 dependencies  | 32 sec    | 20 sec                              |
+
+The shortest build time (the limit) is appriximate 15 seconds and requires no limitation of the build job resources and minimum number of dependencies that are pulled in during the build phase.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:40]:
```suggestion
The shortest build time (the limit) is approximately 15 seconds and requires no limitation of the build job resources and a minimum number of dependencies that are pulled in during the build phase.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+Running multiple function build jobs at once ( especially such with no limits) may drain the cluster resources. To mitigate such risk there is additional limit of 5 symultanous function builds. If a sixth one will be scheduled, it will be built once there will be a vacancy in the build queue.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:42]:
```suggestion
Running multiple Function build jobs at once (especially with no limits) may drain the cluster resources. To mitigate such risk, there is an additional limit of 5 simultaneous Function builds. If a sixth one is scheduled, it is built once a vacancy is in the build queue.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+This limitation is configurable via [`containers.manager.envs.functionBuildMaxSimultaneousJobs`](../../../05-technical-reference/00-configuration-parameters/svls-01-serverless-chart.md#configurable-parameters)
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:44]:
```suggestion
This limitation is configurable using [`containers.manager.envs.functionBuildMaxSimultaneousJobs`](../../../05-technical-reference/00-configuration-parameters/svls-01-serverless-chart.md#configurable-parameters) 
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+The shortest build time (the limit) is appriximate 15 seconds and requires no limitation of the build job resources and minimum number of dependencies that are pulled in during the build phase.
+
+Running multiple function build jobs at once ( especially such with no limits) may drain the cluster resources. To mitigate such risk there is additional limit of 5 symultanous function builds. If a sixth one will be scheduled, it will be built once there will be a vacancy in the build queue.
+
+This limitation is configurable via [`containers.manager.envs.functionBuildMaxSimultaneousJobs`](../../../05-technical-reference/00-configuration-parameters/svls-01-serverless-chart.md#configurable-parameters) 
+
+
+## Runtime phase limitations
+In the runtime the functions are serving user provided logic wrapped in WEB framework (`express` for Node.js and `bottle` for Python). Taking the user logic aside, those frameworks have limitation by themselves and depend on the selected [runtime profile](../../../05-technical-reference/svls-09-available-presets.md#functions-resources) and kubernetes nodes specification (see the reference specification at the end of this article).
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:48]:
```suggestion
In the runtime, the Functions serve user-provided logic wrapped in the WEB framework (`express` for Node.js and `bottle` for Python). Taking the user logic aside, those frameworks have limitations and depend on the selected [runtime profile](../../../05-technical-reference/svls-09-available-presets.md#functions-resources) and the Kubernetes nodes specification (see the note with reference specification at the end of this article).
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+The following describes response times of a selected runtime profiles for a "hello world" function requested at 50 requests/second. This describes the overhead of the serving framework itself. Any user logic added on top of that will ofcourse add extra miliseconds and needs to be profiled separately.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:50]:
```suggestion
The following describes the response times of selected runtime profiles for a "hello world" Function requested at 50 requests/second. This describes the overhead of the serving framework itself. Any user logic added on top of that will add extra milliseconds and must be profiled separately.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+#### Node.js functions
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:52]:
```suggestion
#### Node.js Functions
```
plus see my previous comment on tables

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+|                               | XL     | L      | M      | S      | XS      |
+|-------------------------------|--------|--------|--------|--------|---------|
+| response time [avarage]       | ~13ms  | 13ms   | ~15ms  | ~60ms  | ~400ms  |
+| response time [95 percentile] | ~20ms  | ~30ms  | ~70ms  | ~200ms | ~800ms  |
+| response time [99 percentile] | ~200ms | ~200ms | ~220ms | ~500ms | ~1.25ms |
+
+#### Python functions
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:60]:
```suggestion
#### Python Functions
```
see my previous comment on tables

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+|                               | XL     | L      | M      | S      |
+|-------------------------------|--------|--------|--------|--------|
+| response time [avarage]       | ~11ms  | 12ms   | ~12ms  | ~14ms  |
+| response time [95 percentile] | ~25ms  | ~25ms  | ~25ms  | ~25ms  |
+| response time [99 percentile] | ~175ms | ~180ms | ~210ms | ~280ms |
+
+
+Obviously, the bigger runtime profile, the more resources are available to serve response quicker. Please consider those limits of the serving layer as a baseline - as this do not take your function logic into account.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:69]:
```suggestion
Obviously, the bigger the runtime profile, the more resources are available to serve the response quicker. Please consider those limits of the serving layer as a baseline - as this does not take your Function logic into account.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+
+### Scaling
+
+Function runtime pods can be scaled horizontally from zero up to the limits of the available resources at the kubernetes worker nodes.
+Please find the guide [here](../../../03-tutorials/00-serverless/svls-15-use-external-scalers.md).
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:75]:
```suggestion
See the [Use external scalers](../..///03-tutorials/00-serverless/svls-15-use-external-scalers.md) tutorial for more information.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+## In-cluster docker registry limitations
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:77]:
```suggestion
## In-cluster Docker registry limitations
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+Serverless comes with in-cluster docker registry for function images.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:79]:
```suggestion
Serverless comes with an in-cluster docker registry for the Function images.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+Because of it's [limitations](../../main-areas/serverless/svls-03-container-registries.md) this registry is only suitable for development:
+ - registry capacity is limited to 20GB
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:81]:
```suggestion
 - Registry capacity is limited to 20GB
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+ - there is no image lifecycle managment. Once an image gets stored in the registry it stays there until they are manually removed.
+
+ >> Reference specification of the cluster used for profiling: 
+ All measurements were done on a kubernetes with 5 AWS nodes of type `m5.xlarge worker` (4 CPU 3.1 Ghz x86_64 cores, 16 GiB memory)
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:85]:
```suggestion

```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+| response time [99 percentile] | ~175ms | ~180ms | ~210ms | ~280ms |
+
+
+Obviously, the bigger the runtime profile, the more resources are available to serve the response quicker. Please consider those limits of the serving layer as a baseline - as this does not take your Function logic into account.
+
+
+### Scaling
+
+Function runtime Pods can be scaled horizontally from zero up to the limits of the available resources at the the Kubernetes worker nodes.
+See the [Use external scalers](../../03-tutorials/00-serverless/svls-15-use-external-scalers.md) tutorial for more information.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:75]:
```suggestion
See the [Use external scalers](../../../03-tutorials/00-serverless/svls-15-use-external-scalers.md) tutorial for more information.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+Refer to the [architecture](../../../05-technical-reference/00-architecture/svls-01-architecture.md) diagram for more details.
+
+Having this in mind serverless controller does not have a requirement for horizontal scaling.
+It scales vertically up to the `160Mi` of memory and `500m` of cpu time.
+
+## Limitation for amount of functions
+There is no upper limit of functions that can be run on kyma (similar to kubernetes workloads in general). Once user defines a function, its build jobs and runtime pods will be always requested by serverless controller but it's up to kubernetes to schedule them based on the available memory and cpu time on the k8s worker nodes. This is determined mainly by amount of k8s worker nodes (and the node auto-scaling capabilities) and  their computational capacity.
+
+## Build phase limitation:
+As described before, the build-time phase requires cpu-time and memory to schedule jobs.
```
[the remark this answers is not in the document]
Comment (kwiatekus) [reply docs/01-overview/main-areas/serverless/svls-07-limitations.md]: done

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 16]
```diff
@@ -0,0 +1,103 @@
+---
+title: Serverless limitations
+---
+
+## Controller limitations
+Serverless controller does not serve time-critical requests from users.
+It reconciles Function custom resources (CR), stored at the Kubernetes API Server, and has no persistent state on its own.
+
+Serverless controller doesn't build or serve functions using its allocated runtime resources. It delegates this work to the dedicated Kubernetes workloads. It schedules (build-time) jobs to build the Function Docker image and (runtime) Pods to serve them once they are built.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:9]:
```suggestion
Serverless controller doesn't build or serve Functions using its allocated runtime resources. It delegates this work to the dedicated Kubernetes workloads. It schedules (build-time) jobs to build the Function Docker image and (runtime) Pods to serve them once they are built. 
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+Having this in mind Serverless controller does not require horizontal scaling.
+It scales vertically up to the `160Mi` of memory and `500m` of CPU time.
+
+## Limitation for the number of Functions
+There is no upper limit of Functions that can be run on Kyma (similar to Kubernetes workloads in general). Once a user defines a Function, its build jobs and runtime Pods will always be requested by Serverless controller. It's up to Kubernetes to schedule them based on the available memory and CPU time on the Kubernetes worker nodes. This is determined mainly by the number of the Kubernetes worker nodes (and the node auto-scaling capabilities) and their computational capacity.
+
+## Build phase limitation:
+The time necessary to build Function depends on:
+ - selected [build profile](../../../05-technical-reference/svls-09-available-presets.md#build-jobs-resources) that determines the requested resources ( and their limits ) for the build phase
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:20]:
```suggestion
 - selected [build profile](../../../05-technical-reference/svls-09-available-presets.md#build-jobs-resources) that determines the requested resources (and their limits) for the build phase 
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+|-----------------|-----------|-------------------------------------|
+| no dependencies | 30 sec    | 16 sec                              |
+| 2 dependencies  | 32 sec    | 20 sec                              |
+
+  </details>
+</div>
+
+The shortest build time (the limit) is approximately 15 seconds and requires no limitation of the build job resources and a minimum number of dependencies that are pulled in during the build phase.
+
+Running multiple Function build jobs at once (especially with no limits) may drain the cluster resources. To mitigate such risk, there is an additional limit of 5 simultaneous Function builds. If a sixth one is scheduled, it is built once a vacancy is in the build queue.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:51]:
```suggestion
Running multiple Function build jobs at once (especially with no limits) may drain the cluster resources. To mitigate such risk, there is an additional limit of 5 simultaneous Function builds. If a sixth one is scheduled, it is built once there is a vacancy in the build queue.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+This limitation is configurable using [`containers.manager.envs.functionBuildMaxSimultaneousJobs`](../../../05-technical-reference/00-configuration-parameters/svls-01-serverless-chart.md#configurable-parameters)
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:53]:
```suggestion
This limitation is configurable using [`containers.manager.envs.functionBuildMaxSimultaneousJobs`](../../../05-technical-reference/00-configuration-parameters/svls-01-serverless-chart.md#configurable-parameters).
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+
+This limitation is configurable using [`containers.manager.envs.functionBuildMaxSimultaneousJobs`](../../../05-technical-reference/00-configuration-parameters/svls-01-serverless-chart.md#configurable-parameters) 
+
+
+## Runtime phase limitations
+In the runtime, the Functions serve user-provided logic wrapped in the WEB framework (`express` for Node.js and `bottle` for Python). Taking the user logic aside, those frameworks have limitations and depend on the selected [runtime profile](../../../05-technical-reference/svls-09-available-presets.md#functions-resources) and the Kubernetes nodes specification (see the note with reference specification at the end of this article).
+
+The following describes the response times of selected runtime profiles for a "hello world" Function requested at 50 requests/second. This describes the overhead of the serving framework itself. Any user logic added on top of that will add extra milliseconds and must be profiled separately.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:59]:
```suggestion
The following describes the response times of the selected runtime profiles for a "hello world" Function requested at 50 requests/second. This describes the overhead of the serving framework itself. Any user logic added on top of that will add extra milliseconds and must be profiled separately.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+|                               | XL     | L      | M      | S      |
+|-------------------------------|--------|--------|--------|--------|
+| response time [avarage]       | ~11ms  | 12ms   | ~12ms  | ~14ms  |
+| response time [95 percentile] | ~25ms  | ~25ms  | ~25ms  | ~25ms  |
+| response time [99 percentile] | ~175ms | ~180ms | ~210ms | ~280ms |
+  </details>
+</div>
+
+Obviously, the bigger the runtime profile, the more resources are available to serve the response quicker. Please consider those limits of the serving layer as a baseline - as this does not take your Function logic into account.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:88]:
```suggestion
Obviously, the bigger the runtime profile, the more resources are available to serve the response quicker. Consider these limits of the serving layer as a baseline - as this does not take your Function logic into account.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+
+### Scaling
+
+Function runtime Pods can be scaled horizontally from zero up to the limits of the available resources at the the Kubernetes worker nodes.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:93]:
```suggestion
Function runtime Pods can be scaled horizontally from zero up to the limits of the available resources at the Kubernetes worker nodes.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+See the [Use external scalers](../../../03-tutorials/00-serverless/svls-15-use-external-scalers.md) tutorial for more information.
+
+## In-cluster Docker registry limitations
+
+Serverless comes with an in-cluster docker registry for the Function images.
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:98]:
```suggestion
Serverless comes with an in-cluster Docker registry for the Function images.
```

File: docs/01-overview/main-areas/serverless/svls-07-limitations.md [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+This registry is only suitable for development because of its [limitations](../../main-areas/serverless/svls-03-container-registries.md), i.e.:
+ - Registry capacity is limited to 20GB
+ - There is no image lifecycle management. Once an image is stored in the registry, it stays there until it is manually removed.
+
+> **NOTE:** All measurements were done on Kubernetes with five AWS worker nodes of type `m5.xlarge` (four CPU 3.1 GHz x86_64 cores, 16 GiB memory)
```
Comment (grego952) [inline docs/01-overview/main-areas/serverless/svls-07-limitations.md:103]:
```suggestion
> **NOTE:** All measurements were done on Kubernetes with five AWS worker nodes of type `m5.xlarge` (four CPU 3.1 GHz x86_64 cores, 16 GiB memory).
```

**Pull request merged**